adapter/repository/in-memory-simple/src/main/java/org/mobilitydata/gtfsvalidator/db/InMemoryGtfsDataRepository.java

```diff
@@ -30,6 +32,7 @@
 public class InMemoryGtfsDataRepository implements GtfsDataRepository {
     private final Map<String, Agency> agencyCollection = new HashMap<>();
     private final Map<String, Route> routeCollection = new HashMap<>();
+    private final Map<String, Map<LocalDateTime, CalendarDate>> calendarDateCollection = new HashMap<>();
```

Comment:
A comment describing the different parts of this declaration would be welcomed

Reply:
This map won't work as written because `service_id` isn't guaranteed to be unique in calendar_dates.txt.

[calendar_dates.txt](http://gtfs.org/reference/static#calendar_datestxt) is used in two ways in GTFS:
1. It defines exceptions to service defined in calendar.txt. So you may have weekdays that normally run service_id 1, but if a holiday Dec 25th falls on a Monday, you'd have 2 records here - one saying remove service_id 1 on Dec 25th, and a second one saying add service_id 2 on Dec 25th (for whatever the Dec 25th service is defined by service_id 2). In this case service_id isn't a primary key for records in calendar_dates.txt (it's a foreign key to calendar.txt). The combination of all 3 fields (service_id, date, and exception_type) would be a unique within calendar_dates.txt, if you're looking for a composite key.
2. If agencies don't use calendar.txt to define service, they can provide a massive dump of dates in calendar_dates.txt that says which service_id is running on which date. In this case, `service_id` is a primary key for calendar_dates.txt and all service will be exception_type `1` for added.

So in case number 1 above, `service_id` can't be used as the key in the Map, because it's not unique in calendar_dates.txt.

One option that would work across both cases is if you create a composite key of the 3 fields and use that as the key in the `Map`, and have just a single level map like `Map<String, CalendarDate> calendarDateCollection`.

Reply:
Thanks you for you insight, will implement the solution you suggest.